README.md
HTTP Rewrite Tool based on hrt tool
=================================================

epry是前端代理工具，根据配置把指定的URL指向到本地文件或远程URL。

### 安装

```
npm install prt -g
```

### 使用方法

1. 修改浏览器代理设置，IP：`127.0.0.1`，端口：`2222`，推荐用SwitchySharp（Chrome插件）、FoxyProxy（Firefox插件）切换代理。

2. 创建配置文件 `epry-config.js` ，添加跳转规则。

	代理文件：
	```js
	exports.map = [
		['http://localhost:8080/statics/javascript/lib.js', 'D:\\project\\statics\\javascript\\lib.js']
	];
	```

	代理目录：
	```js
	exports.map = [
		['http://localhost:8080/statics', 'D:\\project\\statics']
	];
	```

3. 在命令行输入 `epry epry-config.js` ，启动HTTP服务。

	```bash
	# 修改端口
	epry epry-config.js --port=2222,默认端口是2222
	# 输出调试信息
	epry epry-config.js --debug=true
	```

### 高级用法

1. 移除版本号。
	```js
	exports.before = function(url) {
		return url.replace(/([^?]+)_\d+(\.(?:js|css))/, '$1$2');
	};
	```

2. 修改文件内容。
	```js
	exports.merge = function(path, callback) {
		// 所有JS头部添加注释
		if (/\.js$/.test(path)) {
			var content = Util.readFileSync(path, 'utf-8');
			return callback('application/javascript', '/* test /*\n' + content);
		}
		// 其它请求
		var contentType = require('mime').lookup(path);
		var buffer = this.util.readFileSync(path);
		return callback(contentType, buffer);
	};
	```
	注：当配置文件里有 `exports.merge` 时会接管所有请求，所以在程序逻辑里需要加入文件类型判断。

3. 修改URL内容。
	```js
	exports.merge = function(path, callback) {
		if (/^http:\/\/(www|wwwtest|beta)\.everbridge\.com\/statics\/javascripts\//.test(this.req.url)) {
			this.util.get(this.req.url, 'gbk', function(body) {
				callback('text/html', body + '<!-- test -->', 'gbk');
			});
			return;
		}
		// 其它请求
		var contentType = require('mime').lookup(path);
		var buffer = this.util.readFileSync(path);
		return callback(contentType, buffer);
	};
	```

changelog.txt
#######################################################################
#
# PRT 变更记录
#
#######################################################################


test/util-test.js

var test = require('tap').test;

var util = require('../util');

test('util.rewrite (empty map)', function(t) {
	var map = [];

	var url = 'http://js.tudouui.com/js/lib/tuilib2_10.js';
	var result = util.rewrite(map, url);
	t.equal(result, 'http://js.tudouui.com/js/lib/tuilib2_10.js');

	t.end();
});

test('util.rewrite (to local file)', function(t) {
	var root = 'D:\\project\\kindeditor';
	var localPath = require('path').resolve(root + '/kindeditor.js');

	var map = [
		['http://js.tudouui.com/js/lib/kindeditor-min.js', localPath],
	];

	var url = 'http://js.tudouui.com/js/lib/kindeditor-min.js';
	var result = util.rewrite(map, url);
	t.equal(result, 'D:\\project\\kindeditor\\kindeditor.js');

	t.end();
});

test('util.rewrite (to local file: partial URL)', function(t) {
	// 1
	var root = 'D:\\project';
	var localPath = require('path').resolve(root + '/kindeditor/');
	var map = [
		['http://js.tudouui.com/js/lib', localPath],
	];
	var url = 'http://js.tudouui.com/js/lib/kindeditor.js';
	var result = util.rewrite(map, url);
	t.equal(result, 'D:\\project\\kindeditor\\kindeditor.js');

	// 2
	var map = [
		['http://www.kindsoft.net/ke4', 'D:\\project\\kindsoft\\ke4'],
	];
	var url = 'http://www.kindsoft.net/ke4/kindeditor.js';
	var result = util.rewrite(map, url);
	t.equal(result, 'D:\\project\\kindsoft\\ke4\\kindeditor.js');

	// 3
	var url = 'http://www.kindsoft.net/ke4/themes/default/default.css?t=20121118.css';
	var result = util.rewrite(map, url);
	t.equal(result, 'D:\\project\\kindsoft\\ke4\\themes\\default\\default.css');

	// 4
	var url = 'http://www.kindsoft.net/';
	var result = util.rewrite(map, url);
	t.equal(result, 'http://www.kindsoft.net/');

	t.end();
});

test('util.rewrite (to remote file: partial URL)', function(t) {
	// 1
	var map = [
		['http://js.tudouui.com/js/lib', 'http://jstest.tudouui.com/js/lib'],
	];
	var url = 'http://js.tudouui.com/js/lib/tuilib2_10.js';
	var result = util.rewrite(map, url);
	t.equal(result, 'http://jstest.tudouui.com/js/lib/tuilib2_10.js');

	// 2
	var map = [
		['http://js.tudouui.com/js/lib/tuilib2_10.js', 'http://jstest.tudouui.com/js/lib/tuilib2_10.js'],
	];
	var url = 'http://js.tudouui.com/js/lib/tuilib2_10.js';
	var result = util.rewrite(map, url);
	t.equal(result, 'http://jstest.tudouui.com/js/lib/tuilib2_10.js');

	// 3
	var url = 'http://js.tudouui.com/js/lib/tuilib2_10.js?t=20121118.css';
	var result = util.rewrite(map, url);
	t.equal(result, 'http://jstest.tudouui.com/js/lib/tuilib2_10.js?t=20121118.css');

	// 4
	var url = 'http://www.kindsoft.net/';
	var result = util.rewrite(map, url);
	t.equal(result, 'http://www.kindsoft.net/');

	t.end();
});

test('util.rewrite (rewrite all)', function(t) {
	var root = 'D:\\project\\kindeditor';
	var localPath = require('path').resolve(root + '/kindeditor.js');

	// 1
	var map = [
		['http://js.tudouui.com/js/lib/kindeditor-min.js', localPath],
	];

	var url = '/js/lib/kindeditor-min.js';
	var result = util.rewrite(map, url, root);
	t.equal(result, 'D:\\project\\kindeditor\\kindeditor.js');

	// 2
	var map = [
		['http://js.tudouui.com/js/lib/kindeditor-min.js', localPath],
	];

	var url = '/js/lib/kindeditor-all.js';
	var result = util.rewrite(map, url, root);
	t.equal(result, 'D:\\project\\kindeditor\\js\\lib\\kindeditor-all.js');

	t.end();
});

util.js

var Path = require('path');
var Fs = require('fs');
var Iconv = require('iconv-lite');
var Request = require('request');

function each(obj, fn) {
	for (var key in obj) {

		if (obj.hasOwnProperty(key)) {
			if (fn.call(obj[key], key, obj[key]) === false) {
				break;
			}
		}

	}
}

function isRegExp(val) {
	return Object.prototype.toString.call(val) === '[object RegExp]';
}

function undef(val, defaultVal) {
	return typeof val === 'undefined' ? defaultVal : val;
}

function info(str) {
	console.info('\033[36m', str, '\033[0m');
}

function warn(str) {
	console.warn('\033[33m', str, '\033[0m');
}

function error(str) {
	console.error('\033[31m', str, '\033[0m');
}

function readFileSync(filePath, encoding) {
	var buffer = new Buffer('');

	try {
		buffer = Fs.readFileSync(filePath);
	} catch (e) {
		error(e.toString());
	}

	if (!encoding) {
		return buffer;
	}

	var fileStr = Iconv.fromEncoding(buffer, encoding);

	return fileStr;
}

function get(url, encoding, callback) {
	Request({
		url : url,
		encoding : null
	}, function (error, response, buffer) {
		if (!error && response.statusCode == 200) {
			if (encoding) {
				buffer = Iconv.fromEncoding(buffer, encoding);
			}
			callback(buffer);
		}
	});
}

function loadPlugin(name) {
	return require(__dirname + '/plugins/' + name + '.js');
}

// return value: url or path
function rewrite(map, url, serverRoot) {
	// rewrite by map
	for (var i = 0, len = map.length; i < len; i++) {
		var row = map[i];

		if (row.length != 2) {
			continue;
		}

		var from = row[0];
		var to = row[1];

		if (serverRoot) {
			from = from.replace(/^https?:\/\/[^\/]+/, '');
		}

		var index = url.indexOf(from);

		if (index === 0) {
			var suffix = url.substr(index + from.length);

			if (/^https?:\/\//.test(to)) {
				return to + suffix;
			}

			suffix = suffix.replace(/[?#].*$/, '');
			return Path.resolve(to + suffix);
		}
	}
	// rewrite all
	if (serverRoot) {
		var to = serverRoot + url.replace(/^https?:\/\/[^\/]+|\?.*$/, '');
		return Path.resolve(to);
	}
	return url;
}

exports.each = each;
exports.isRegExp = isRegExp;
exports.undef = undef;
exports.info = info;
exports.warn = warn;
exports.error = error;
exports.loadPlugin = loadPlugin;
exports.readFileSync = readFileSync;
exports.get = get;
exports.rewrite = rewrite;


plugins/mall.js
var Fs = require('fs');
var Path = require('path');
var Util = require('../util');
var Less = require('less');
var Mime = require('mime');
var Iconv = require('iconv-lite');
var ReactTools = require('react-tools');

var RE_AUTOFIXNAME = /define\((?=[^'"])/;

// URL移除版本号
function stripVersionInfo(url) {
	return url.replace(/([^?]+)_\d+(\.(?:js|css|swf|png|jpg|gif))/i, '$1$2');
}

// CSS扩展名改成LESS
function cssToLess(url) {
	return url.replace(/\.css$/, '.less');
}

function resolveUrl(url) {
	while(true) {
		url = url.replace(/\w+\/\.\.\//g, '');
		if (!/\.\.\//.test(url)) {
			break;
		}
	}
	url = url.replace(/\.\//g, '');
	return url;
}

// 将JS代码改成AMD模块，包含路径转换，补充模块ID，模板转换等
function fixModule(path, str) {
	var root = path.replace(/^(.*?)[\\\/](src|build|dist)[\\\/].*$/, '$1');
	var relativePath = path.split(Path.sep).join('/').replace(/^.+\/src\//, '');
	var mid = relativePath.replace(/\.js$/, '');
	function fixDep(s, format) {
		if (format) {
			s = s.replace(/\s/g, '');
		}
		return s.replace(/(['"])(.+?)\1(,?)/g, function($0, $1, $2, $3) {
			var f = $2;
			if(f.charAt(0) == '.') {
				f = relativePath.replace(/[\w-]+\.js$/, '') + f;
				f = resolveUrl(f);
			}
			else if(f.charAt(0) == '/') {
				f = f.slice(1);
			}
			if (format) {
				return '\n  "' + f + '"' + $3 + '\n';
			} else {
				return $1 + f + $1 + $3;
			}
		}).replace(/,\n\n/g, ',\n');
	}

	// 补充模块ID
	if(/(?:^|[^\w\.])define\s*\(/.test(str) && !/(?:^|[^\w\.])define\s*\(\s*['"]/.test(str)) {
		str = str.replace(/\b(define\s*\(\s*)/, '$1"' + mid + '", ');
	}

	// 补齐依赖
	str = str.replace(/((?:^|[^\w\.])define\s*\(\s*['"].*?['"]\s*,\s*)([['"][\s\S]+?)(,\s*function\s*\()/g, function($0, $1, $2, $3) {
		return $1 + fixDep($2, true) + $3;
	});
	str = str.replace(/((?:^|[^\w\.])require\s*\(\s*)([\['"][\s\S]+?)(,\s*function\s*\()/g, function($0, $1, $2, $3) {
		return $1 + fixDep($2, false) + $3;
	});
	str = str.replace(/((?:^|[^\w\.])define\s*\(\s*['"].*?['"]\s*)(,\s*function\s*\()/g, '$1,[]$2');

	// 非AMD模块
	if(!/(?:^|[^\w\.])(define|require)\s*\(/.test(str)) {
		return str += '\n/* autogeneration */\n"define" in this && define("' + mid + '", [], function(){});\n';
	}

	return str;
}

// Replace require.text to string
function replaceTemplate(path, str) {
	var root = path.replace(/^(.*?)[\\\/](src|build|dist)[\\\/].*$/, '$1');
	// sub template
	function replaceSubTemplate(parentPath, str) {
		Util.info('[import template] ' + parentPath);
		return str.replace(/<%\s*require\.text\(\s*(['"])(.+?)\1\s*\);?\s*%>/g, function($0, $1, $2) {
			var f = $2;
			if(/^[a-z_/]/i.test(f)) {
				f = root + '/src/' + f;
			}
			else {
				f = parentPath.replace(/[\w-]+\.\w+$/, '') + f;
				f = resolveUrl(f);
			}
			var s = Util.readFileSync(f, 'utf-8');
			s = replaceSubTemplate(f, s);
			s = s.replace(/^\uFEFF/, '');
			return s;
		});
	}

	// replace template string
	str = str.replace(/(\b)require\.text\(\s*(['"])(.+?)\2\s*\)/g, function($0, $1, $2, $3) {
		var f = $3;
		if(/^[a-z_/]/i.test(f)) {
			f = root + '/src/' + f;
		}
		else {
			f = path.replace(/[\w-]+\.\w+$/, '') + f;
			f = resolveUrl(f);
		}
		var s = Util.readFileSync(f, 'utf-8');
		s = replaceSubTemplate(f, s);
		s = s.replace(/^\uFEFF/, '');
		s = s.replace(/\\/g, '\\\\');
		s = s.replace(/(\r\n|\r|\n)\s*/g, '\\n');
		s = s.replace(/'/g, "\\'");
		return $1 + "'" + s + "'";
	});

	return str;
}

// 合并本地文件
function merge(path, callback) {
	var root = path.replace(/^(.*?)[\\\/](src|build|dist)[\\\/].*$/, '$1');

	var newPath = path.split(Path.sep).join('/');

	// CSS
	if (/\.less$/.test(newPath)) {
		var str = Util.readFileSync(path, 'utf-8');

		var parser = new(Less.Parser)({
			env : 'development',
			dumpLineNumbers : 'comments',
			paths : ['.', root + '/src'],
			filename : path,
		});

		parser.parse(str, function(error, tree) {
			if (error) {
				return console.error(error);
			}
			callback('text/css', tree.toCSS());
		});
		return;
	}

	if (!/src(\/[^\/]+)+\/(lib|lite|loader|react)\.js$/.test(newPath) && /src(\/[^\/]+)+\/.+\.js$/.test(newPath)) {
		if(Fs.existsSync(path)){
			var str = Util.readFileSync(path, 'utf-8');
			str = fixModule(path, str);
			str = replaceTemplate(path, str);
			if (str.indexOf('use strict') > 0) {
				str = ReactTools.transform(str, { harmony: true });
				// console.log(ReactTools.transform(str));
			}
			return callback('application/javascript', str);
		}else{
			return console.error('Not find File: '+ path)
		}
	}

	var contentType = Mime.lookup(path);
	var buffer = Util.readFileSync(path);

	return callback(contentType, buffer);
}

exports.stripVersionInfo = stripVersionInfo;
exports.cssToLess = cssToLess;
exports.merge = merge;


plugins/merge.js
var Fs = require('fs');
var Path = require('path');
var Util = require('../util');
var Less = require('less');
var Mime = require('mime');
var Iconv = require('iconv-lite');
var ReactTools = require('react-tools');

var RE_AUTOFIXNAME = /define\((?=[^'"])/;

// URL移除版本号
function stripVersionInfo(url) {
	return url.replace(/([^?]+)_\d+(\.(?:js|css|swf|png|jpg|gif))/i, '$1$2');
}

// CSS扩展名改成LESS
function cssToLess(url) {
	return url.replace(/\.css/, '.less');
}

function resolveUrl(url) {
	while(true) {
		url = url.replace(/\w+\/\.\.\//g, '');
		if (!/\.\.\//.test(url)) {
			break;
		}
	}
	url = url.replace(/\.\//g, '');
	return url;
}

// 将JS代码改成AMD模块，包含路径转换，补充模块ID，模板转换等
function fixModule(path, str) {
	var root = path.replace(/^(.*?)[\\\/](src|build|dist)[\\\/].*$/, '$1');
	var relativePath = path.split(Path.sep).join('/').replace(/^.+\/src\/js\//, '');
	var mid = relativePath.replace(/\.js$/, '');

	function fixDep(s, format) {
		if (format) {
			s = s.replace(/\s/g, '');
		}
		return s.replace(/(['"])(.+?)\1(,?)/g, function($0, $1, $2, $3) {
			var f = $2;
			if(f.charAt(0) == '.') {
				f = relativePath.replace(/[\w-]+\.js$/, '') + f;
				f = resolveUrl(f);
			}
			else if(f.charAt(0) == '/') {
				f = f.slice(1);
			}
			if (format) {
				return '\n  "' + f + '"' + $3 + '\n';
			} else {
				return $1 + f + $1 + $3;
			}
		}).replace(/,\n\n/g, ',\n');
	}

	// 补充模块ID
	if(/(?:^|[^\w\.])define\s*\(/.test(str) && !/(?:^|[^\w\.])define\s*\(\s*['"]/.test(str)) {
		str = str.replace(/\b(define\s*\(\s*)/, '$1"' + mid + '", ');
	}

	// 补齐依赖
	str = str.replace(/((?:^|[^\w\.])define\s*\(\s*['"].*?['"]\s*,\s*)([['"][\s\S]+?)(,\s*function\s*\()/g, function($0, $1, $2, $3) {
		return $1 + fixDep($2, true) + $3;
	});
	str = str.replace(/((?:^|[^\w\.])require\s*\(\s*)([\['"][\s\S]+?)(,\s*function\s*\()/g, function($0, $1, $2, $3) {
		return $1 + fixDep($2, false) + $3;
	});
	str = str.replace(/((?:^|[^\w\.])define\s*\(\s*['"].*?['"]\s*)(,\s*function\s*\()/g, '$1,[]$2');

	// 非AMD模块
	if(!/(?:^|[^\w\.])(define|require)\s*\(/.test(str)) {
		return str += '\n/* autogeneration */\n"define" in this && define("' + mid + '", [], function(){});\n';
	}

	return str;
}

function replaceCombomText(path,str){
	var sourceMapRoot=path.replace(/^(.*?)[\\\/](javascripts)[\\\/].*$/, '$1');
	var subDir = /js\//.test(path)?"js":'javascripts';
	var regSourceMapExp=/\/\*\s*@sourceMap:(.|\s)+?;\s*\*\//ig;///\/\*.*sourceMap:.*;+?\*\//ig;
	var sourceMap=regSourceMapExp.exec(str);
	var dist='';
	//获取sourceMap的list
	if(sourceMap&&sourceMap[0]){
		var sourceMapList=sourceMap[0].split("@sourceMap:");
		for(var i=0;i<sourceMapList.length;i++){
			if((/\.js/g).test(sourceMapList[i])){
				var sourceMapUrl=sourceMapList[i].match(/.+?\.js/g);
				var sourcePath=sourceMapRoot+'/'+subDir+'/'+sourceMapUrl[0];
				var sourceMapFileStr=Util.readFileSync(sourcePath,'utf8');
				dist+=sourceMapFileStr + '\n';
				//pathMap[sourcePath]=sourceMapFileStr;
			}
		}
		return dist;

	}
	return str;

}

// 合并本地文件
function merge(path, callback) {
	var root = path.replace(/^(.*?)[\\\/](src|build|dist)[\\\/].*$/, '$1');

	var newPath = path.split(Path.sep).join('/');

	// CSS
	if (/\.less/.test(newPath)) {
		var str = Util.readFileSync(path, 'utf-8');

		var parser = new(Less.Parser)({
			env : 'development',
			dumpLineNumbers : 'comments',
			paths : ['.', root + '/src/stylesheet'],
			filename : path,
		});

		parser.parse(str, function(error, tree) {
			if (error) {
				return console.error(error);
			}
			callback('text/css', tree.toCSS());
		});
		return;
	}

	if (!/src\/javascripts\/(lib|lite|loader|react)\.js$/.test(newPath) && /src\/javascripts\/.+\.js$/.test(newPath)) {
		if(Fs.existsSync(path)){
			var str = Util.readFileSync(path, 'utf-8');
			//str = fixModule(path, str);
			str = replaceCombomText(path,str);
			if (str.indexOf('use strict') > 0) {
				str = ReactTools.transform(str, { harmony: true });
				// console.log(ReactTools.transform(str));
			}
			return callback('application/javascript', str);
		}else{
			return console.error('Not find File: '+ path)
		}
	}

	var contentType = Mime.lookup(path);
	var buffer = Util.readFileSync(path);

	return callback(contentType, buffer);
}

exports.stripVersionInfo = stripVersionInfo;
exports.cssToLess = cssToLess;
exports.merge = merge;


config/default.js
'use strict';
//exports.serverRoot = 'D:\\program\\webapp\\src';
//exports.serverRoot = '/Users/dengyuzhe/WebstormProjects/youku';
exports.serverRoot = '/Users/dengyuzhe/WebstormProjects/webapp';

exports.map = [
	//线上环境

	//wwwtest环境
	//['http://js.ykimg.com/youku/dist', exports.serverRoot+'/src/'],
	//['http://css.ykimg.com/youku/dist', exports.serverRoot+'/src/'],
	//['http://localhost:8080/statics', localHost+'src/'],
	//['http://10.24.32.112:8181/statics', localHost+'src/'],
	//['http://http://10.24.32.156:8080/statics', localHost+'src/'],
	//['https://manager-qa1.everbridge.net/statics', localHost+'src/']

	//['http://localhost:8080/statics', exports.serverRoot+'/src'],
	//['http:///10.24.32.156:8080/statics', exports.serverRoot+'/src'],
	['http://10.24.32.156:8080/statics', exports.serverRoot+'/src']

	//['http://localhost:8080/statics/javascripts', exports.serverRoot+'\\javascripts'],
	//['http://localhost:8080/statics/javascripts', exports.serverRoot+'\\javascripts'],
	//['http://localhost:8080/statics/javascripts', exports.serverRoot+'\\javascripts'],
	//['http://10.24.32.112:8181/statics/javascripts', exports.serverRoot+'\\javascripts'],
	//['https://manager-qa1.everbridge.net/statics/javascripts', exports.serverRoot+'\\javascripts']

];

exports.before = function(url) {
	var Merge = this.util.loadPlugin('merge');

	url = Merge.stripVersionInfo(url);
	url = Merge.cssToLess(url);
	return url;
};

exports.merge = function(path, callback) {
	var Merge = this.util.loadPlugin('merge');

	Merge.merge.call(this, path, callback);
};


main.js

var Http = require('http');
var Path = require('path');
var Fs = require('fs');
var Iconv = require('iconv-lite');
var Url = require('url');
var Mime = require('mime');
var Request = require('request');
var Watchr = require('watchr');

var HttpProxy = require('http-proxy');
var Util = require('./util');

var Optimist = require('optimist');

Optimist.usage([
	'Usage: pry /path/to/config.js --port=[number] --debug=[true/false]\n\n',
	'Examples:\n',
	'pry ./config/my.js\n',
	'pry ./config/my.js --port=8080\n',
	'pry ./config/my.js --debug=true',
].join(''));

var ARGV = Optimist.argv;

if (ARGV.help || ARGV.h) {
	Optimist.showHelp();
	process.exit();
}

if (ARGV.version || ARGV.v) {
	var packageInfo = JSON.parse(Fs.readFileSync(__dirname + '/package.json', 'utf-8'));
	console.log(packageInfo.version);
	process.exit();
}

var PORT = Util.undef(ARGV.port, 2222);
var CONFIG_FILE = ARGV._.length > 0 ? ARGV._[0] : __dirname + '/config/default.js';
var DEBUG = Util.undef(ARGV.debug, false);

CONFIG_FILE = Path.resolve(CONFIG_FILE);

var CONFIG = require(CONFIG_FILE);

function setResponse(response, contentType, buffer) {
	response.setHeader('Content-Type', contentType);
	response.setHeader('Access-Control-Allow-Origin', '*');
	response.write(buffer);
	response.end();
}

function main() {
	// reload config
	Watchr.watch({
		paths: [CONFIG_FILE],
		listeners: {
			error: function(err) {
				Util.error('[watch] ', err);
			},
			change: function() {
				delete require.cache[CONFIG_FILE];
				CONFIG = require(CONFIG_FILE);
				Util.info('[watch] reload config: ' + CONFIG_FILE);
			}
		}
	});

	// start server
	Http.createServer(function(request, response) {

		var url = request.url;
		var before = CONFIG.before;
		var map = CONFIG.map;
		var merge = CONFIG.merge;

		var me = {
			config : CONFIG,
			util : Util,
			req : request,
			res : response,
		}

		if (DEBUG) {
			console.log('[get] ' + url);
		}

		var from = url;

		if (before) {
			from = before.call(me, from);
		}

		if (from) {
			var serverRoot = /^https?:\/\//.test(url) ? '' : CONFIG.serverRoot;

			var to = Util.rewrite(map, from, serverRoot);

			// rewrite
			if(from !== to){
				Util.info('[rewrite] ' + url + ' -> ' + to);

				// local file
				if (!/^https?:\/\//.test(to)) {
					if (merge) {
						merge.call(me, to, function(contentType, buffer, encoding) {
							if (typeof buffer == 'string' && encoding) {
								buffer = Iconv.toEncoding(buffer, encoding);
							}
							setResponse(response, contentType, buffer);

						});
						return;
					}

					var contentType = Mime.lookup(to);
					var buffer = Util.readFileSync(to);

					setResponse(response, contentType, buffer);
					return;
				}

				// remote URL
				request.pipe(Request(to)).pipe(response);
				return;
			}
		}

		// no rewrite
		var parsed = Url.parse(url);

		var proxy = new HttpProxy.createProxyServer({
			target : {
				host : parsed.hostname,
				port : parsed.port || 80,
			},
		});
		proxy.on('error', function (err, req, res) {
			// res.writeHead(500, {'Content-Type': 'text/plain'});
			if ('/favicon.ico' != parsed.path) { // favicon.ico 出错不显示
				Util.error('[proxy] '+ req.url);
			}

			res.end('Something went wrong. And we are reporting a custom error message.');
		});


		proxy.web(request, response);

	}).listen(PORT);

	console.log('Rewrite Server runing at port: ' + PORT);
}

main();
